File: app/src/main/java/com/drdisagree/colorblendr/utils/FabricatedUtil.java
package com.drdisagree.colorblendr.utils;

import static com.drdisagree.colorblendr.common.Const.FABRICATED_OVERLAY_NAME_APPS;
import static com.drdisagree.colorblendr.common.Const.MONET_BACKGROUND_LIGHTNESS;
import static com.drdisagree.colorblendr.common.Const.MONET_PITCH_BLACK_THEME;
import static com.drdisagree.colorblendr.common.Const.TINT_TEXT_COLOR;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.Color;

import androidx.annotation.ColorInt;
import androidx.core.util.Pair;

import com.drdisagree.colorblendr.common.Const;
import com.drdisagree.colorblendr.config.RPrefs;
import com.drdisagree.colorblendr.utils.fabricated.FabricatedOverlayResource;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class FabricatedUtil {

    public static void createDynamicOverlay(
            FabricatedOverlayResource overlay,
            ArrayList<ArrayList<Integer>> paletteLight,
            ArrayList<ArrayList<Integer>> paletteDark
    ) {
        assignDynamicPaletteToOverlay(overlay, true /* isDark */, paletteDark);
        assignDynamicPaletteToOverlay(overlay, false /* isDark */, paletteLight);
        assignFixedColorsToOverlay(overlay, paletteLight);
    }

    @SuppressWarnings("unchecked")
    private static void assignDynamicPaletteToOverlay(FabricatedOverlayResource overlay, boolean isDark, ArrayList<ArrayList<Integer>> palette) {
        String suffix = isDark ? "dark" : "light";
        boolean pitchBlackTheme = RPrefs.getBoolean(MONET_PITCH_BLACK_THEME, false);
        DynamicColors.ALL_DYNAMIC_COLORS_MAPPED.forEach(pair -> {
            String resourceName = "system_" + pair.first + "_" + suffix;
            int colorValue;

            Pair<Object, Object> valPair = (Pair<Object, Object>) pair.second;
            if (valPair.first instanceof String) {
                colorValue = Color.parseColor((String) (isDark ?
                        valPair.second :
                        valPair.first
                ));
            } else {
                Pair<Integer, Integer> colorIndexPair = (Pair<Integer, Integer>) valPair.second;
                colorValue = palette.get((Integer) valPair.first).get(
                        isDark ?
                                colorIndexPair.second :
                                colorIndexPair.first
                );

                colorValue = replaceColorForPitchBlackTheme(
                        pitchBlackTheme,
                        resourceName,
                        colorValue,
                        isDark ?
                                colorIndexPair.second :
                                colorIndexPair.first
                );
            }

            overlay.setColor(resourceName, colorValue);
        });
    }

    private static void assignFixedColorsToOverlay(FabricatedOverlayResource overlay, ArrayList<ArrayList<Integer>> paletteLight) {
        DynamicColors.FIXED_COLORS_MAPPED.forEach(pair -> {
            String resourceName = "system_" + pair.first;
            int colorValue = paletteLight.get(pair.second.first).get(pair.second.second);
            overlay.setColor(resourceName, colorValue);
        });
    }

    public static void assignPerAppColorsToOverlay(FabricatedOverlayResource overlay, ArrayList<ArrayList<Integer>> palette) {
        DynamicColors.M3_REF_PALETTE.forEach(pair -> {
            // Format of pair: <resourceName, <lightnessToChange, <colorIndexRow, colorIndexColumn>>>

            boolean pitchBlackTheme = RPrefs.getBoolean(MONET_PITCH_BLACK_THEME, false);

            String resourceName = pair.first;

            Pair<Integer, Pair<Integer, Integer>> valPair = pair.second;

            // TODO: Use lightnessToChange to modify the color value
            // int lightnessToChange = valPair.first + 100;

            Pair<Integer, Integer> colorIndexPair = valPair.second;
            int baseColor = palette.get(colorIndexPair.first).get(colorIndexPair.second);
            baseColor = replaceColorForPitchBlackTheme(pitchBlackTheme, resourceName, baseColor, colorIndexPair.second);

            overlay.setColor(resourceName, baseColor);
            overlay.setColor("g" + resourceName, baseColor);
        });

        if (!RPrefs.getBoolean(TINT_TEXT_COLOR, true)) {
            addTintlessTextColors(overlay);
        }
    }

    public static void getAndSaveSelectedFabricatedApps(Context context) {
        PackageManager packageManager = context.getPackageManager();
        List<ApplicationInfo> applications = packageManager.getInstalledApplications(PackageManager.GET_META_DATA);
        HashMap<String, Boolean> selectedApps = new HashMap<>();

        for (ApplicationInfo appInfo : applications) {
            String packageName = appInfo.packageName;
            boolean isSelected = OverlayManager.isOverlayEnabled(
                    String.format(FABRICATED_OVERLAY_NAME_APPS, packageName)
            );

            if (isSelected) {
                selectedApps.put(packageName, true);
            }
        }

        Const.saveSelectedFabricatedApps(selectedApps);
    }

    public static @ColorInt int replaceColorForPitchBlackTheme(boolean pitchBlackTheme, String resourceName, int colorValue, int colorIndex) {
        if (pitchBlackTheme) {
            int lightness = RPrefs.getInt(MONET_BACKGROUND_LIGHTNESS, 100);
            return switch (resourceName) {
                case "m3_ref_palette_dynamic_neutral_variant6",
                        "gm3_ref_palette_dynamic_neutral_variant6",
                        "system_background_dark",
                        "system_surface_dark" -> Color.BLACK;
                case "m3_ref_palette_dynamic_neutral_variant12",
                        "gm3_ref_palette_dynamic_neutral_variant12" ->
                        ColorUtil.modifyLightness(colorValue, lightness - 40, colorIndex);
                case "m3_ref_palette_dynamic_neutral_variant17",
                        "gm3_ref_palette_dynamic_neutral_variant17" ->
                        ColorUtil.modifyLightness(colorValue, lightness - 60, colorIndex);
                case "system_surface_container_dark" ->
                        ColorUtil.modifyLightness(colorValue, lightness - 20, colorIndex);
                default -> colorValue;
            };
        }

        return colorValue;
    }

    public static void addTintlessTextColors(FabricatedOverlayResource overlay) {
        String[] prefixes = new String[]{
                "m3_sys_color_",
                "m3_sys_color_dynamic_"
        };
        String[] variants = new String[]{
                "dark_",
                "light_",
        };
        String[] suffixes = new String[]{
                "on_surface",
                "on_surface_variant",
                "on_background"
        };

        for (String prefix : prefixes) {
            for (String variant : variants) {
                for (String suffix : suffixes) {
                    String resourceName = prefix + variant + suffix;
                    overlay.setColor(
                            resourceName,
                            variant.contains("dark") ?
                                    Color.WHITE :
                                    Color.BLACK
                    );
                }
            }
        }

        // Dark mode
        ArrayList<Pair<String, Integer>> resourcesDark = new ArrayList<>();
        resourcesDark.add(new Pair<>("m3_ref_palette_dynamic_neutral90", Color.WHITE));
        resourcesDark.add(new Pair<>("m3_ref_palette_dynamic_neutral95", Color.WHITE));
        resourcesDark.add(new Pair<>("m3_ref_palette_dynamic_neutral_variant70", 0xB3FFFFFF));
        resourcesDark.add(new Pair<>("m3_ref_palette_dynamic_neutral_variant80", Color.WHITE));
        resourcesDark.add(new Pair<>("text_color_primary_dark", Color.WHITE));
        resourcesDark.add(new Pair<>("text_color_secondary_dark", 0xB3FFFFFF));
        resourcesDark.add(new Pair<>("text_color_tertiary_dark", 0x80FFFFFF));

        // Light mode
        ArrayList<Pair<String, Integer>> resourcesLight = new ArrayList<>();
        resourcesLight.add(new Pair<>("m3_ref_palette_dynamic_neutral10", Color.BLACK));
        resourcesLight.add(new Pair<>("m3_ref_palette_dynamic_neutral_variant30", 0xB3000000));
        resourcesLight.add(new Pair<>("text_color_primary_light", Color.BLACK));
        resourcesLight.add(new Pair<>("text_color_secondary_light", 0xB3000000));
        resourcesLight.add(new Pair<>("text_color_tertiary_light", 0x80000000));

        for (Pair<String, Integer> pair : resourcesDark) {
            overlay.setColor(pair.first, pair.second);
            overlay.setColor("g" + pair.first, pair.second);
        }

        for (Pair<String, Integer> pair : resourcesLight) {
            overlay.setColor(pair.first, pair.second);
            overlay.setColor("g" + pair.first, pair.second);
        }
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/utils/FabricatedUtil.kt
package com.drdisagree.colorblendr.utils

import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import androidx.annotation.ColorInt
import androidx.core.util.component1
import androidx.core.util.component2
import com.drdisagree.colorblendr.data.common.Const
import com.drdisagree.colorblendr.data.common.Const.FABRICATED_OVERLAY_NAME_APPS
import com.drdisagree.colorblendr.data.common.Const.MONET_PITCH_BLACK_THEME
import com.drdisagree.colorblendr.data.common.Const.THEMING_ENABLED
import com.drdisagree.colorblendr.data.common.Const.TINT_TEXT_COLOR
import com.drdisagree.colorblendr.data.common.Const.saveSelectedFabricatedApps
import com.drdisagree.colorblendr.data.common.Const.workingMethod
import com.drdisagree.colorblendr.data.config.Prefs.getBoolean
import com.drdisagree.colorblendr.utils.ColorUtil.adjustLightness
import com.drdisagree.colorblendr.utils.ColorUtil.getColorNamesM3
import com.drdisagree.colorblendr.utils.DynamicColors.ALL_DYNAMIC_COLORS_MAPPED
import com.drdisagree.colorblendr.utils.DynamicColors.FIXED_COLORS_MAPPED
import com.drdisagree.colorblendr.utils.DynamicColors.M3_REF_PALETTE
import com.drdisagree.colorblendr.utils.fabricated.FabricatedOverlayResource

object FabricatedUtil {

    private val colorNamesM3Variants = listOf(
        getColorNamesM3(isDynamic = false, prefixG = false),
        getColorNamesM3(isDynamic = true, prefixG = false),
        getColorNamesM3(isDynamic = true, prefixG = true),
        getColorNamesM3(isDynamic = false, prefixG = true)
    )

    fun FabricatedOverlayResource.createDynamicOverlay(
        paletteLight: ArrayList<ArrayList<Int>>,
        paletteDark: ArrayList<ArrayList<Int>>
    ) {
        assignDynamicPaletteToOverlay(true,  /* isDark */paletteDark)
        assignDynamicPaletteToOverlay(false,  /* isDark */paletteLight)
        assignFixedColorsToOverlay(paletteLight)
    }

    private fun FabricatedOverlayResource.assignDynamicPaletteToOverlay(
        isDark: Boolean,
        palette: ArrayList<ArrayList<Int>>
    ) {
        val suffix = if (isDark) "dark" else "light"
        val isPitchBlackTheme = getBoolean(MONET_PITCH_BLACK_THEME, false)
        val prefixSuffix = arrayOf(
            "system_" to "_${suffix}",
            "m3_sys_color_${suffix}_" to "",
            "m3_sys_color_dynamic_${suffix}_" to "",
            "gm3_sys_color_${suffix}_" to "",
            "gm3_sys_color_dynamic_${suffix}_" to ""
        )

        ALL_DYNAMIC_COLORS_MAPPED.forEach { colorMapping ->
            for ((tempPrefix, tempSuffix) in prefixSuffix) {
                val (resourceName, colorValue) = colorMapping.extractResourceFromColorMap(
                    prefix = tempPrefix,
                    suffix = tempSuffix,
                    palette = palette,
                    isDark = isDark
                ).let { (name, value) ->
                    name to applyColorAdjustments(
                        colorMapping,
                        name,
                        value,
                        isDark,
                        isPitchBlackTheme
                    )
                }

                setColor(resourceName, colorValue)
            }
        }
    }

    private fun FabricatedOverlayResource.assignFixedColorsToOverlay(
        paletteLight: ArrayList<ArrayList<Int>>
    ) {
        FIXED_COLORS_MAPPED.forEach { colorMapping ->
            val (resourceName, colorValue) = colorMapping.extractResourceFromColorMap(
                prefix = "system_",
                palette = paletteLight
            )

            setColor(resourceName, colorValue)
        }
    }

    fun FabricatedOverlayResource.assignPerAppColorsToOverlay(
        palette: ArrayList<ArrayList<Int>>
    ) {
        val isPitchBlackTheme = getBoolean(MONET_PITCH_BLACK_THEME, false)
        val tintTextColor = getBoolean(TINT_TEXT_COLOR, true)

        M3_REF_PALETTE.forEach { colorMapping ->
            val (resourceName, colorValue) = colorMapping.extractResourceFromColorMap(
                palette = palette
            ).let { (name, value) ->
                name to applyColorAdjustments(
                    colorMapping,
                    name,
                    value,
                    isDark = false,
                    isPitchBlackTheme
                )
            }

            setColor(resourceName, colorValue)
            setColor("g$resourceName", colorValue)
        }

        colorNamesM3Variants.forEach { variant ->
            variant.forEachIndexed { i, row ->
                row.forEachIndexed { j, name ->
                    setColor(name, palette[i][j])
                }
            }
        }

        replaceColorsPerPackageName(palette, isPitchBlackTheme)

        if (!tintTextColor) {
            addTintlessTextColors()
        }
    }

    fun updateFabricatedAppList(context: Context) {
        if (workingMethod != Const.WorkMethod.ROOT ||
            !getBoolean(THEMING_ENABLED, true)
        ) return

        val packageManager = context.packageManager
        val applications = packageManager.getInstalledApplications(PackageManager.GET_META_DATA)

        val selectedApps = HashMap<String, Boolean>().apply {
            applications.forEach { appInfo ->
                val packageName = appInfo.packageName
                val isSelected = OverlayManager.isOverlayEnabled(
                    String.format(FABRICATED_OVERLAY_NAME_APPS, packageName)
                )

                if (isSelected) {
                    put(packageName, true)
                }
            }
        }

        //        selectedApps.put(BuildConfig.APPLICATION_ID, true);

        saveSelectedFabricatedApps(selectedApps)
    }

    private fun applyColorAdjustments(
        colorMapping: ColorMapping,
        resourceName: String,
        colorValue: Int,
        isDark: Boolean,
        pitchBlackTheme: Boolean
    ): Int {
        return adjustColorForPitchBlackThemeIfRequired(
            pitchBlackTheme,
            resourceName,
            colorValue
        ).let { adjustedValue ->
            colorMapping.adjustColorBrightnessIfRequired(adjustedValue, isDark)
        }
    }

    @ColorInt
    fun adjustColorForPitchBlackThemeIfRequired(
        pitchBlackTheme: Boolean,
        resourceName: String,
        colorValue: Int
    ): Int {
        if (!pitchBlackTheme) return colorValue

        return when (resourceName) {
            "m3_ref_palette_dynamic_neutral_variant6",
            "gm3_ref_palette_dynamic_neutral_variant6",
            "system_background_dark",
            "system_surface_dark" -> {
                Color.BLACK
            }

            "m3_ref_palette_dynamic_neutral_variant12",
            "gm3_ref_palette_dynamic_neutral_variant12" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -56)
            }

            "m3_ref_palette_dynamic_neutral_variant17",
            "gm3_ref_palette_dynamic_neutral_variant17",
            "gm3_system_bar_color_night" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -74)
            }

            "system_surface_container_lowest_dark" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -54)
            }

            "system_surface_container_low_dark" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -45)
            }

            "system_surface_container_dark" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -36)
            }

            "system_surface_container_high_dark",
            "system_surface_dim_dark" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -25)
            }

            "system_surface_container_highest_dark",
            "system_surface_bright_dark" -> {
                adjustLightness(color = colorValue, brightnessPercentage = -16)
            }

            else -> {
                colorValue
            }
        }
    }

    private fun FabricatedOverlayResource.addTintlessTextColors() {
        val prefixes = arrayOf("m3_sys_color_", "m3_sys_color_dynamic_")
        val variants = arrayOf("dark_", "light_")
        val suffixes = arrayOf("on_surface", "on_surface_variant", "on_background")

        prefixes.forEach { prefix ->
            variants.forEach { variant ->
                suffixes.forEach { suffix ->
                    setColor(
                        "$prefix$variant$suffix",
                        if (variant.contains("dark")) Color.WHITE else Color.BLACK
                    )
                }
            }
        }

        // Resources for dark and light modes
        val resources = mapOf(
            "dark" to listOf(
                "m3_ref_palette_dynamic_neutral90" to Color.WHITE,
                "m3_ref_palette_dynamic_neutral95" to Color.WHITE,
                "m3_ref_palette_dynamic_neutral_variant70" to -0x333334,
                "m3_ref_palette_dynamic_neutral_variant80" to Color.WHITE,
                "text_color_primary_dark" to Color.WHITE,
                "text_color_secondary_dark" to -0x4c000001,
                "text_color_tertiary_dark" to -0x7f000001,
                "google_dark_default_color_on_background" to Color.WHITE,
                "gm_ref_palette_grey500" to Color.WHITE
            ),
            "light" to listOf(
                "m3_ref_palette_dynamic_neutral10" to Color.BLACK,
                "m3_ref_palette_dynamic_neutral_variant30" to -0x4d000000,
                "text_color_primary_light" to Color.BLACK,
                "text_color_secondary_light" to -0x4d000000,
                "text_color_tertiary_light" to -0x80000000,
                "google_default_color_on_background" to Color.BLACK,
                "gm_ref_palette_grey700" to Color.BLACK
            )
        )

        resources.forEach { (_, pairs) ->
            pairs.forEach { (name, color) ->
                setColor(name, color)
                setColor("g$name", color)
            }
        }

        when {
            targetPackage == "com.android.settings" -> {
                when {
                    Build.VERSION.SDK_INT <= Build.VERSION_CODES.UPSIDE_DOWN_CAKE -> {
                        setColor(
                            "settingslib_text_color_primary_device_default",
                            Color.WHITE,
                            "night"
                        )
                        setColor(
                            "settingslib_text_color_secondary_device_default",
                            -0x4c000001,
                            "night"
                        )
                    }

                    Build.VERSION.SDK_INT > Build.VERSION_CODES.UPSIDE_DOWN_CAKE -> {
                        setColor(
                            "settingslib_materialColorOnSurface",
                            Color.WHITE,
                            "night"
                        )
                        setColor(
                            "settingslib_materialColorOnSurfaceVariant",
                            -0x4c000001,
                            "night"
                        )
                    }
                }
            }
        }
    }
}
